nil pointer dereference in extractHeadlessEndpoints

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!
-->

**What happened**:

 I recently started observing this crash:

```
...
time="2021-03-26T21:33:22Z" level=debug msg="No endpoints could be generated from service external-dns/external-dns"
time="2021-03-26T21:33:22Z" level=debug msg="No endpoints could be generated from service podinfo/podinfo"
time="2021-03-26T21:33:22Z" level=debug msg="No endpoints could be generated from service home-assistant/home-assistant-codeserver"
time="2021-03-26T21:33:22Z" level=debug msg="No endpoints could be generated from service monitoring/kube-prometheus-stack-grafana"
time="2021-03-26T21:33:22Z" level=debug msg="No endpoints could be generated from service kube-system/kube-dns"
time="2021-03-26T21:33:22Z" level=debug msg="Generating matching endpoint alertmanager-operated.mrv.thebends.org with EndpointAddress IP 192.168.173.8"
time="2021-03-26T21:33:22Z" level=debug msg="Generating matching endpoint alertmanager-kube-prometheus-stack-alertmanager-0.alertmanager-operated.mrv.thebends.org with EndpointAddress IP 192.168.173.8"
time="2021-03-26T21:33:22Z" level=debug msg="Endpoints generated from service: monitoring/alertmanager-operated: [alertmanager-kube-prometheus-stack-alertmanager-0.alertmanager-operated.mrv.thebends.org 0 IN A  192.168.173.8 [] alertmanager-operated.mrv.thebends.org 0 IN A  192.168.173.8 []]"
time="2021-03-26T21:33:22Z" level=debug msg="No endpoints could be generated from service monitoring/kube-prometheus-stack-alertmanager"
time="2021-03-26T21:33:22Z" level=debug msg="No endpoints could be generated from service monitoring/speedtest-prometheus"
time="2021-03-26T21:33:22Z" level=debug msg="No endpoints could be generated from service home-assistant/home-assistant-postgresql"
time="2021-03-26T21:33:22Z" level=debug msg="No endpoints could be generated from service kubernetes-dashboard/kubernetes-dashboard"
time="2021-03-26T21:33:22Z" level=debug msg="No endpoints could be generated from service pihole/pihole-web"
time="2021-03-26T21:33:22Z" level=debug msg="No endpoints could be generated from service cert-manager/cert-manager"
time="2021-03-26T21:33:22Z" level=debug msg="No endpoints could be generated from service monitoring/kube-prometheus-stack-prometheus-node-exporter"
time="2021-03-26T21:33:22Z" level=debug msg="No endpoints could be generated from service monitoring/kube-prometheus-stack-kube-state-metrics"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x19d4062]
goroutine 1 [running]:
sigs.k8s.io/external-dns/source.(*serviceSource).extractHeadlessEndpoints(0xc0006aed20, 0xc00083a630, 0xc000b53a80, 0x30, 0x0, 0xc000804b70, 0xc0006dd178, 0xc0006dd0e0)
	/bitnami/blacksmith-sandox/external-dns-0.7.6/src/github.com/kubernetes-incubator/external-dns/source/service.go:287 +0x582
sigs.k8s.io/external-dns/source.(*serviceSource).generateEndpoints(0xc0006aed20, 0xc00083a630, 0xc000b53a80, 0x31, 0x3cd7730, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/bitnami/blacksmith-sandox/external-dns-0.7.6/src/github.com/kubernetes-incubator/external-dns/source/service.go:481 +0xb93
sigs.k8s.io/external-dns/source.(*serviceSource).endpointsFromTemplate(0xc0006aed20, 0xc00083a630, 0x0, 0x0, 0x0, 0x0, 0x2)
	/bitnami/blacksmith-sandox/external-dns-0.7.6/src/github.com/kubernetes-incubator/external-dns/source/service.go:365 +0x2b9
sigs.k8s.io/external-dns/source.(*serviceSource).Endpoints(0xc0006aed20, 0x2b13760, 0xc000b4db00, 0x4, 0x4, 0x33, 0x5c, 0xc00014e2a0)
	/bitnami/blacksmith-sandox/external-dns-0.7.6/src/github.com/kubernetes-incubator/external-dns/source/service.go:195 +0x7e5
sigs.k8s.io/external-dns/source.(*multiSource).Endpoints(0xc000b29ec0, 0x2b13760, 0xc000b4db00, 0x8, 0x3cd9320, 0x1, 0x203000, 0xc00065f130)
	/bitnami/blacksmith-sandox/external-dns-0.7.6/src/github.com/kubernetes-incubator/external-dns/source/multi_source.go:35 +0xc5
sigs.k8s.io/external-dns/source.(*dedupSource).Endpoints(0xc0003a0f30, 0x2b13760, 0xc000b4db00, 0x3c78f70, 0x222b1c0, 0xc000b48c80, 0x2b13760, 0xc000b4db00)
	/bitnami/blacksmith-sandox/external-dns-0.7.6/src/github.com/kubernetes-incubator/external-dns/source/dedup_source.go:42 +0xca
sigs.k8s.io/external-dns/controller.(*Controller).RunOnce(0xc0006aef00, 0x2b13760, 0xc000b4db00, 0x3ca5b20, 0x25e7801)
	/bitnami/blacksmith-sandox/external-dns-0.7.6/src/github.com/kubernetes-incubator/external-dns/controller/controller.go:136 +0x16e
sigs.k8s.io/external-dns/controller.(*Controller).Run(0xc0006aef00, 0x2b136a0, 0xc000463400)
	/bitnami/blacksmith-sandox/external-dns-0.7.6/src/github.com/kubernetes-incubator/external-dns/controller/controller.go:194 +0x20f
main.main()
	/bitnami/blacksmith-sandox/external-dns-0.7.6/src/github.com/kubernetes-incubator/external-dns/main.go:354 +0xe50
```

**What you expected to happen**:

I would expect an error message for an invalid cluster state, rather than a crash.

**How to reproduce it (as minimally and precisely as possible)**:

I have not yet been able to determine which service is causing the problem.

**Anything else we need to know?**:

**Environment**:
- External-DNS version (use `external-dns --version`): 0.7.6
- DNS provider:  `rfc2136`
- Others: Running the helm chart https://github.com/bitnami/charts/tree/master/bitnami/external-dns - Observed in 4.9.3 and  4.9.2

--- a/source/service.go
+++ b/source/service.go
@@ -284,7 +284,7 @@ func (sc *serviceSource) extractHeadlessEndpoints(svc *v1.Service, hostname stri
 
 		for _, address := range addresses {
 			// find pod for this address
-			if address.TargetRef.APIVersion != "" || address.TargetRef.Kind != "Pod" {
+			if address.TargetRef == nil || address.TargetRef.APIVersion != "" || address.TargetRef.Kind != "Pod" {
 				log.Debugf("Skipping address because its target is not a pod: %v", address)
 				continue
 			}

--- a/source/service_test.go
+++ b/source/service_test.go
@@ -2423,6 +2423,7 @@ func TestHeadlessServicesHostIP(t *testing.T) {
 		podnames                 []string
 		hostnames                []string
 		podsReady                []bool
+		targetRefs               []*v1.ObjectReference
 		publishNotReadyAddresses bool
 		expected                 []*endpoint.Endpoint
 		expectError              bool
@@ -2449,6 +2450,10 @@ func TestHeadlessServicesHostIP(t *testing.T) {
 			[]string{"foo-0", "foo-1"},
 			[]string{"foo-0", "foo-1"},
 			[]bool{true, true},
+			[]*v1.ObjectReference{
+				{APIVersion: "", Kind: "Pod", Name: "foo-0"},
+				{APIVersion: "", Kind: "Pod", Name: "foo-1"},
+			},
 			false,
 			[]*endpoint.Endpoint{
 				{DNSName: "foo-0.service.example.org", Targets: endpoint.Targets{"1.1.1.1"}},
@@ -2479,6 +2484,10 @@ func TestHeadlessServicesHostIP(t *testing.T) {
 			[]string{"foo-0", "foo-1"},
 			[]string{"foo-0", "foo-1"},
 			[]bool{true, true},
+			[]*v1.ObjectReference{
+				{APIVersion: "", Kind: "Pod", Name: "foo-0"},
+				{APIVersion: "", Kind: "Pod", Name: "foo-1"},
+			},
 			false,
 			[]*endpoint.Endpoint{},
 			false,
@@ -2506,6 +2515,10 @@ func TestHeadlessServicesHostIP(t *testing.T) {
 			[]string{"foo-0", "foo-1"},
 			[]string{"foo-0", "foo-1"},
 			[]bool{true, true},
+			[]*v1.ObjectReference{
+				{APIVersion: "", Kind: "Pod", Name: "foo-0"},
+				{APIVersion: "", Kind: "Pod", Name: "foo-1"},
+			},
 			false,
 			[]*endpoint.Endpoint{
 				{DNSName: "foo-0.service.example.org", Targets: endpoint.Targets{"1.1.1.1"}, RecordTTL: endpoint.TTL(1)},
@@ -2536,6 +2549,10 @@ func TestHeadlessServicesHostIP(t *testing.T) {
 			[]string{"foo-0", "foo-1"},
 			[]string{"foo-0", "foo-1"},
 			[]bool{true, false},
+			[]*v1.ObjectReference{
+				{APIVersion: "", Kind: "Pod", Name: "foo-0"},
+				{APIVersion: "", Kind: "Pod", Name: "foo-1"},
+			},
 			false,
 			[]*endpoint.Endpoint{
 				{DNSName: "foo-0.service.example.org", Targets: endpoint.Targets{"1.1.1.1"}},
@@ -2565,6 +2582,10 @@ func TestHeadlessServicesHostIP(t *testing.T) {
 			[]string{"foo-0", "foo-1"},
 			[]string{"foo-0", "foo-1"},
 			[]bool{true, false},
+			[]*v1.ObjectReference{
+				{APIVersion: "", Kind: "Pod", Name: "foo-0"},
+				{APIVersion: "", Kind: "Pod", Name: "foo-1"},
+			},
 			true,
 			[]*endpoint.Endpoint{
 				{DNSName: "foo-0.service.example.org", Targets: endpoint.Targets{"1.1.1.1"}},
@@ -2595,10 +2616,41 @@ func TestHeadlessServicesHostIP(t *testing.T) {
 			[]string{"foo-0", "foo-1"},
 			[]string{"", ""},
 			[]bool{true, true},
+			[]*v1.ObjectReference{
+				{APIVersion: "", Kind: "Pod", Name: "foo-0"},
+				{APIVersion: "", Kind: "Pod", Name: "foo-1"},
+			},
 			false,
 			[]*endpoint.Endpoint{
 				{DNSName: "service.example.org", Targets: endpoint.Targets{"1.1.1.1", "1.1.1.2"}},
 			},
+			false,
+		},
+		{
+			"annotated Headless services without a targetRef has no endpoints",
+			"",
+			"testing",
+			"foo",
+			v1.ServiceTypeClusterIP,
+			"",
+			"",
+			false,
+			map[string]string{"component": "foo"},
+			map[string]string{
+				hostnameAnnotationKey: "service.example.org",
+			},
+			v1.ClusterIPNone,
+			[]string{"1.1.1.1"},
+			map[string]string{
+				"component": "foo",
+			},
+			[]string{},
+			[]string{"foo-0"},
+			[]string{"foo-0"},
+			[]bool{true, true},
+			[]*v1.ObjectReference{nil},
+			false,
+			[]*endpoint.Endpoint{},
 			false,
 		},
 	} {
@@ -2648,11 +2700,7 @@ func TestHeadlessServicesHostIP(t *testing.T) {
 
 				address := v1.EndpointAddress{
 					IP: "4.3.2.1",
-					TargetRef: &v1.ObjectReference{
-						APIVersion: "",
-						Kind:       "Pod",
-						Name:       podname,
-					},
+					TargetRef: tc.targetRefs[i],
 				}
 				if tc.podsReady[i] {
 					addresses = append(addresses, address)

